two_factor/forms.py

```diff
@@ -152,6 +160,22 @@ def __init__(self, user, initial_device, **kwargs):
                 label=label
             )
 
+    def clean_device_id(self):
+        if self.data.get("device_id"):
+            try:
+                for user_device in devices_for_user(self.user):
+                    if user_device.persistent_id == self.data["device_id"]:
+                        self.device_cache = user_device
+                        break
+            except ObjectDoesNotExist:
+                raise forms.ValidationError(self.error_messages['invalid_device_id'])
+
+    def _chosen_device(self, user):
```

Comment:
This method doesn't appear to be called by anything, what's it for?

Reply:
`_chosen_device` is the method called by the `OTPTokenForm._clean_otp` method in django-otp.

It currently returns None which makes django-otp try to match the token against each user's devices in order until one works (in `match_token`) but since `match_token` uses `devices_for_token` that iterates over the devices that way `for model in device_classes():` there is no guarantee that it won't try the backup devices first.

I have added more details in my comment at the related issue at : https://github.com/jazzband/django-two-factor-auth/issues/473#issuecomment-1196384536 under the heading "Full investigation".